### PR TITLE
Refactoring and addition of unit tests

### PR DESCRIPTION
**Pull request by nasdas-dev**
This pull request includes new unit tests for three classes which increases their line/statement coverage.
They contribute to issue #6207.

Further, I have also refactored existing unit tests to make them more readable, understandable and maintanable.
In this sense, I focused more on test smells described in the paper Refactoring test code (van Deursen et al.) and also on the Code Howtos (Test Cases) in the Development Documentation.

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/test/java/org/jabref/logic/citationstyle/CitationStyleCacheTest.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  private List<BibEntry> entries;
+  private BibDatabase database;
+  private BibDatabaseContext databaseContext;
+  private CitationStyleCache csCache;
+
+  @Test
+  void getCitationForTest() {
+    BibEntry bibEntry = new BibEntry();
+    bibEntry.setCitationKey("test");
+    List<BibEntry> entries = new ArrayList<>();
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/logic/citationstyle/CitationStyleCacheTest.java:26]: You can use List.of(...)

File: src/test/java/org/jabref/logic/importer/OpenDatabaseTest.java [hunk 2]
```diff
@@ -28,19 +28,24 @@
...
-    private final Path bibEncodingWithoutNewline;
+    private Path bibNoHeader = null;
+    private Path bibWrongHeader = null;
+    private Path bibHeader = null;
+    private Path bibHeaderAndSignature = null;
+    private Path bibEncodingWithoutNewline = null;
     private final FileUpdateMonitor fileMonitor = new DummyFileUpdateMonitor();
 
-    OpenDatabaseTest() throws URISyntaxException {
+    OpenDatabaseTest() {
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/logic/importer/OpenDatabaseTest.java:38]: Why did you change this? If you ignore the tests if the files are not found, we won't see if anyone accidently deleted the files.

File: src/test/java/org/jabref/migrations/SpecialFieldsToSeparateFieldsTest.java [hunk 3]
```diff
@@ -17,10 +18,19 @@
 
 class SpecialFieldsToSeparateFieldsTest {
 
+    private BibEntry entry;
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/migrations/SpecialFieldsToSeparateFieldsTest.java:21]: I see no advantage in using this variant over the existing

File: src/test/java/org/jabref/model/FieldChangeTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class FieldChangeTest {
+
+  private FieldChange Fieldchange;
+
+  @Test
+  void testHashCode() {
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/model/FieldChangeTest.java:18]: Testing for HashCode is useless, as it's a java method

File: src/test/java/org/jabref/model/FieldChangeTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    final String newValueOther = "barX";
+
+    FieldChange fc = new FieldChange(entry, field, oldValue, newValue);
+    FieldChange fcOther = new FieldChange(entryOther, fieldOther, oldValueOther, newValueOther);
+    FieldChange fcBlankAll = new FieldChange(null, null, null, null);
+    FieldChange fcBlankField = new FieldChange(entry, null, oldValue, newValue);
+    FieldChange fcBlankOldValue = new FieldChange(entry, field, null, newValue);
+    FieldChange fcBlankNewValue = new FieldChange(entry, field, oldValue, null);
+
+    assertFalse(fc.equals("foo"));
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/model/FieldChangeTest.java:42]: assertEquals/assertNotEquals  already calls the underlying equals method of the object, so so need to test the equals with true/false

File: src/test/java/org/jabref/model/FieldChangeTest.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    assertFalse(fcBlankAll.equals(fc));
+    assertFalse(fc.equals(fcOther));
+    assertFalse(fcBlankField.equals(fc));
+    assertFalse(fcBlankOldValue.equals(fc));
+    assertFalse(fcBlankNewValue.equals(fc));
+    assertTrue(fcBlankAll.equals(fcBlankAll));
+  }
+
+  @Test
+  void testToString() {
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/model/FieldChangeTest.java:53]: A test for toString is also useless. toString is usually just used internally for development/debugging and might change. No real value in testing this

**Review by Siedlerchr — CHANGES_REQUESTED** [review]

File: src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      AutoCompletePreferences autoCompletePreferences = new AutoCompletePreferences(true, AutoCompleteFirstNameMode.BOTH, AutoCompletePreferences.NameFormat.BOTH, FieldFactory.parseFieldList(personEntryField.getName() + ";" + singleEntryField.getName() + ";" + multipleEntryField.getName() + ";" + journalEntryField.getName() + ";" + publisherEntryField.getName() + ";" + specialEntryField.getName()), null);
+      SuggestionProviders sp = new SuggestionProviders(database, abbreviationRepository, autoCompletePreferences);
+      SuggestionProviders empty = new SuggestionProviders();
+
+      entry.setField(personEntryField, "Goethe");
+      entry.setField(singleEntryField, "Single");
+      entry.setField(multipleEntryField, "Multiple");
+      entry.setField(journalEntryField, "Journal");
+      entry.setField(publisherEntryField, "Publisher");
+      entry.setField(specialEntryField, "2000");
```
Comment (calixtus) [inline src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java:38]:
```
BibEntry entry = new BibEntry()
   .withField(...)
   .withField(...)
...
```

File: src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  void getForFieldTest() {
+      BibDatabase database = new BibDatabase();
+      JournalAbbreviationRepository abbreviationRepository = mock(JournalAbbreviationRepository.class);
+      BibEntry entry = new BibEntry();
+      Field personEntryField = StandardField.AUTHOR;
+      Field singleEntryField = StandardField.XREF;
+      Field multipleEntryField = StandardField.XDATA;
+      Field journalEntryField = StandardField.JOURNAL;
+      Field publisherEntryField = StandardField.PUBLISHER;
+      Field specialEntryField = SpecialField.PRINTED;
```
Comment (calixtus) [inline src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java:28]: Wrapping the StandardFields in another Fields var seems way too 'over-abstracted' to me... StandardField.AUTHOR etc. is imho already a sufficient level of abstraction and I don't see a need to change the personEntryField to another StandardField but AUTHOR, since the test seems comprehensive enough to me.

File: src/test/java/org/jabref/gui/groups/GroupTreeViewModelTest.java [hunk 9]
```diff
@@ -40,12 +40,12 @@ void setUp() throws Exception {
     @Test
     void rootGroupIsAllEntriesByDefault() throws Exception {
         AllEntriesGroup allEntriesGroup = new AllEntriesGroup("All entries");
-        assertEquals(new GroupNodeViewModel(databaseContext, stateManager, taskExecutor, allEntriesGroup, new CustomLocalDragboard(), mock(PreferencesService.class)), groupTree.rootGroupProperty().getValue());
+        assertEquals(new GroupNodeViewModel(databaseContext, stateManager, taskExecutor, allEntriesGroup, new CustomLocalDragboard(), mock(PreferencesService.class)), groupTree.rootGroupProperty().getValue(), "New group view model is not set to all entries by default.");
```
Comment (calixtus) [inline src/test/java/org/jabref/gui/groups/GroupTreeViewModelTest.java:43]: I'm not sure about our test policy @JabRef/developers here? Is it sufficient to express the meaning of the test in the method name or do we additionally output it in the assert clause? What is the suggested standard behavior of junit?

File: src/test/java/org/jabref/gui/journals/JournalAbbreviationsViewModelMixedAbbreviationsTabTest.java [hunk 10]
```diff
@@ -69,8 +69,8 @@ void setUpViewModel(@TempDir Path tempFolder) throws Exception {
 
     @Test
     void testInitialHasNoFilesAndNoAbbreviations() {
-        assertEquals(0, viewModel.journalFilesProperty().size());
-        assertEquals(0, viewModel.abbreviationsProperty().size());
+        assertEquals(0, viewModel.journalFilesProperty().size(), "Incorrect setup of test. Journal files not set to 0 at setup.");
+        assertEquals(0, viewModel.abbreviationsProperty().size(), "Incorrect setup of test. Abbreviations not set to 0 at setup");
```
Comment (calixtus) [inline src/test/java/org/jabref/gui/journals/JournalAbbreviationsViewModelMixedAbbreviationsTabTest.java:73]: Output seems reasonable here, if multiple asserts happen. On the other hand, why are there multiple asserts in one test? Should this be distinguished? Or is this trivial? @JabRef/developers

File: src/test/java/org/jabref/gui/preferences/keybindings/KeyBindingViewModelTest.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+
+class KeyBindingViewModelTest {
+
+  @Test
+  void resetToDefault() {
+    // Set new key binding
+    KeyBindingRepository keyBindingRepository = new KeyBindingRepository();
+    KeyBindingsTabViewModel model = new KeyBindingsTabViewModel(keyBindingRepository, mock(DialogService.class), mock(PreferencesService.class));
```
Comment (calixtus) [inline src/test/java/org/jabref/gui/preferences/keybindings/KeyBindingViewModelTest.java:25]: Var name is confusing, since this is a KeyBindings_TAB_ViewModel like KeyBindingViewModel later. Maybe better `tabViewModel`?

File: src/test/java/org/jabref/logic/importer/OpenDatabaseTest.java (hunk 2 above)
Comment (calixtus) [reply src/test/java/org/jabref/logic/importer/OpenDatabaseTest.java:38]: Agree with @Siedlerchr : This seems to me an intrinsic part of the test: To see, if the files are there.

File: src/test/java/org/jabref/logic/integrity/HTMLCharacterCheckerTest.java [hunk 12]
```diff
@@ -9,12 +9,22 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
 
 public class HTMLCharacterCheckerTest {
 
     private final HTMLCharacterChecker checker = new HTMLCharacterChecker();
     private final BibEntry entry = new BibEntry();
 
+    @Test
+    void fieldNullValueCheck() {
+        Exception exception = assertThrows(
```
Comment (calixtus) [inline src/test/java/org/jabref/logic/integrity/HTMLCharacterCheckerTest.java:21]: Why do you keep the exception var?

File: src/test/java/org/jabref/logic/layout/LayoutEntryTest.java [hunk 13]
```diff
@@ -70,29 +70,29 @@ public String layout(String layoutFile, BibEntry entry) throws IOException {
...
+
+        assertEquals(2, LayoutEntry.parseMethodsCalls("bla(test),foo(fark)").size(), "Parsing of layout method calls failed. Check for arguments in parentheses.");
+        assertEquals("bla", (LayoutEntry.parseMethodsCalls("bla(test),foo(fark)").get(0)).get(0), "Parsing of layout method calls failed. Input did not match expected output. Check for arguments in parentheses.");
+        assertEquals("foo", (LayoutEntry.parseMethodsCalls("bla(test),foo(fark)").get(1)).get(0), "Parsing of layout method calls failed. Input did not match expected output. Check for arguments in parentheses");
+        assertEquals("test", (LayoutEntry.parseMethodsCalls("bla(test),foo(fark)").get(0)).get(1), "Parsing of layout method calls failed. Input did not match expected output. Check for arguments in parentheses");
+        assertEquals("fark", (LayoutEntry.parseMethodsCalls("bla(test),foo(fark)").get(1)).get(1), "Parsing of layout method calls failed. Input did not match expected output. Check for arguments in parentheses");
```
Comment (calixtus) [inline src/test/java/org/jabref/logic/layout/LayoutEntryTest.java:96]: It seems to me that the first sentence "Parsing of layout entry method calls failed." is superfluous, since this is what the method name implies (which is displayed when assert fails). Maybe just remove the first sentence and change the method name to something more meaningful?

File: src/test/java/org/jabref/migrations/SpecialFieldsToSeparateFieldsTest.java [hunk 14]
```diff
@@ -29,41 +39,40 @@ public void migrateToCorrectField(SpecialField field, String fieldInKeyword, Bib
...
     @Test
     public void noKewordToMigrate() {
-        BibEntry entry = new BibEntry().withField(StandardField.AUTHOR, "JabRef")
-                                       .withField(StandardField.KEYWORDS, "tdd");
-        BibEntry expected = new BibEntry().withField(StandardField.AUTHOR, "JabRef")
-                                          .withField(StandardField.KEYWORDS, "tdd");
+        entry.setField(StandardField.AUTHOR, "JabRef");
+        entry.setField(StandardField.KEYWORDS, "tdd");
+        expectedEntry.setField(StandardField.AUTHOR, "JabRef");
+        expectedEntry.setField(StandardField.KEYWORDS, "tdd");
```
Comment (calixtus) [inline src/test/java/org/jabref/migrations/SpecialFieldsToSeparateFieldsTest.java:45]: Please keep creation of the BibEntry in the test method here and use the fluid api to set the fields.

File: src/test/java/org/jabref/logic/journals/AbbreviationTest.java [hunk 15]
```diff
@@ -101,4 +103,18 @@ void testDefaultAndShortestUniqueAbbreviationsAreSame() {
         Abbreviation abbreviation = new Abbreviation("Long Name", "L N");
         assertEquals(abbreviation.getAbbreviation(), abbreviation.getShortestUniqueAbbreviation());
     }
+
+    @Test
+    void testToString() {
+      Abbreviation abbreviation = new Abbreviation("Long Name", "L N", "LN");
+      assertEquals(abbreviation.toString(), "Abbreviation{name=Long Name, abbreviation=L N, medlineAbbreviation=L N, shortestUniqueAbbreviation=LN}");
+    }
```
Comment (calixtus) [inline src/test/java/org/jabref/logic/journals/AbbreviationTest.java:111]: See https://github.com/JabRef/jabref/pull/7597/files#r603937444

File: src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java (hunk 7 above)
Comment (tobiasdiez) [reply src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java:38]: (The purpose of these `with*` methods was actually to use them in lambdas, and not everywhere else...e.g. `entries.map(entry -> entry.withField(...)).collect()`. I don't think chaining them together has much advantages over traditional set calls.)

File: src/test/java/org/jabref/gui/groups/GroupTreeViewModelTest.java (hunk 9 above)
Comment (tobiasdiez) [reply src/test/java/org/jabref/gui/groups/GroupTreeViewModelTest.java:43]: Assert messages should only be used in rare cases and never to explain the purpose of the test (that should be done via the test method name). They should only be used when the junit message is not enough to see where the problem is if the assertion fails. For example, you compare two dates expressed in milliseconds, then junit tells you that 53432423 is not 2323232 - not very helpful. In this case, one can supplement it by saying "date is not first of January 2020".

File: src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java (hunk 7 above)
Comment (koppor) [reply src/test/java/org/jabref/gui/autocompleter/SuggestionProvidersTest.java:38]: These are our builders. See https://github.com/JabRef/jabref/blob/033c706eb1a12b444437ba07408123b9188ce036/src/main/java/org/jabref/model/entry/BibEntry.java#L53.

We don't seem to have time to come up with something more clear such as crafting nice builders or use `@Builder` by Lombock (https://projectlombok.org/features/Builder).

Thus, `withField` is IMHO the way to go.

**Comment by koppor** [comment]
@nasdas-dev May I ask whether you will find time to address our comments?

**Comment by nasdas-dev** [comment]
@koppor I'll have a look this weekend! Sorry for the delay

**Comment by koppor** [comment]
We closed the other PR #7653, so no worries if the files are duplicate.

The normal way would be to do some `git reset` magic, but it's hard to master. I would really recommend to read https://lostechies.com/joshuaflanagan/2010/09/03/use-gitk-to-understand-git/.

Could you work on this PR and resolve the conflicts? Then, we can review and merge.

**Comment by calixtus** [comment]
Hi @nasdas-dev , did you already had some time to get this branch and the corrupt commits right?

File: src/test/java/org/jabref/logic/layout/LayoutEntryTest.java (hunk 13 above)
Comment (koppor) [reply src/test/java/org/jabref/logic/layout/LayoutEntryTest.java:96]: Much work todo, I just reverted the change.

**Comment by koppor** [comment]
@nasdas-dev I took the freedom to finalize the PR. Most importantly, I used the power of JUnit's parameterized tests.

**Review by calixtus — APPROVED** [review]
Looks good to me

**Comment by Siedlerchr** [comment]
one failing unit test
org.jabref.model.FieldChangeTest
```

  Test differentFieldChangeIsNotEqual() FAILED

  org.opentest4j.AssertionFailedError: expected: <FieldChange [entry=, field=DOI, oldValue=foo, newValue=bar]> but was: <FieldChange [entry=, field=DOI, oldValue=fooX, newValue=barX]>
```

**Comment by calixtus** [comment]
> 
> 
> Looks good to me

Way to quick again... I'll look into it.
